test/e2e_node/container_lifecycle_test.go

```diff
@@ -886,6 +887,355 @@ var _ = SIGDescribe(framework.WithSerial(), "Containers Lifecycle", func() {
+			}, f.Timeouts.PodStart, f.Timeouts.Poll).Should(gomega.BeTrueBecause("kubelet should be restarted"))
+
+			ginkgo.By("ensuring that no completed init container is restarted")
+			gomega.Consistently(ctx, func() bool {
+				pod, err = client.Get(ctx, pod.Name, metav1.GetOptions{})
+				framework.ExpectNoError(err)
+				for _, status := range pod.Status.InitContainerStatuses {
+					if status.State.Terminated == nil || status.State.Terminated.ExitCode != 0 {
+						continue
+					}
+
+					if status.RestartCount > 0 {
+						return false
+					}
+				}
+				return true
+			}, 1*time.Minute, f.Timeouts.Poll).Should(gomega.BeTrueBecause("no completed init container should be restarted"))
+
+			ginkgo.By("Analyzing results")
+			// Cannot analyze the results with the container logs as the
+			// container statuses have been removed from container runtime.
+			gomega.Expect(pod.Status.InitContainerStatuses[0].RestartCount).To(gomega.Equal(int32(0)))
+			gomega.Expect(pod.Status.InitContainerStatuses[1].RestartCount).To(gomega.Equal(int32(0)))
+			gomega.Expect(pod.Status.InitContainerStatuses[2].RestartCount).To(gomega.Equal(int32(0)))
+		})
+	})
+
+	ginkgo.When("a Pod is initializing the long-running init container", func() {
+		var client *e2epod.PodClient
+		var err error
+		var pod *v1.Pod
+		init1 := "init-1"
+		init2 := "init-2"
+		longRunningInit3 := "long-running-init-3"
+		regular1 := "regular-1"
+
+		ginkgo.BeforeEach(func(ctx context.Context) {
+			pod = &v1.Pod{
+				ObjectMeta: metav1.ObjectMeta{
+					Name: "initializing-long-running-init-container",
+				},
+				Spec: v1.PodSpec{
+					RestartPolicy: v1.RestartPolicyAlways,
+					InitContainers: []v1.Container{
+						{
+							Name:  init1,
+							Image: busyboxImage,
+							Command: ExecCommand(init1, execCommand{
+								Delay:    1,
+								ExitCode: 0,
+							}),
+						},
+						{
+							Name:  init2,
+							Image: busyboxImage,
+							Command: ExecCommand(init2, execCommand{
+								Delay:    1,
+								ExitCode: 0,
+							}),
+						},
+						{
+							Name:  longRunningInit3,
+							Image: busyboxImage,
+							Command: ExecCommand(longRunningInit3, execCommand{
+								Delay:    300,
+								ExitCode: 0,
+							}),
+						},
+					},
+					Containers: []v1.Container{
+						{
+							Name:  regular1,
+							Image: busyboxImage,
+							Command: ExecCommand(regular1, execCommand{
+								Delay:    300,
+								ExitCode: 0,
+							}),
+						},
+					},
+				},
+			}
+			preparePod(pod)
+
+			client = e2epod.NewPodClient(f)
+			pod = client.Create(ctx, pod)
+			ginkgo.By("Waiting for the pod to be initializing the long-running init container")
+			err := e2epod.WaitForPodCondition(ctx, f.ClientSet, pod.Namespace, pod.Name, "PodInitializing", 1*time.Minute, func(pod *v1.Pod) (bool, error) {
+				for _, c := range pod.Status.InitContainerStatuses {
+					if c.Name != longRunningInit3 {
+						continue
+					}
+					if c.State.Running != nil {
```

Comment:
As with https://github.com/kubernetes/kubernetes/pull/125935#discussion_r1668380613, it might be better to explicitly check if the long-running init container is start running.

```suggestion
					if c.State.Running != nil &&
						(c.Started == nil || *c.Started == true) {
```